### PR TITLE
Binlink stubs

### PR DESCRIPTION
**Pull request by jamessewell**
This PR creates a stub script in /hab/binlinks (can be moved) which is used instead of a direct
link when binlinking.

The scripts look like this:

(file: /hab/binlinks/jirotech/node_exporter/release-0.16/20180926010538/bin/node_exporter)

```
if ! [ -x "$(command -v hab)" ]; then
    echo 'ERROR: The core/hab package needs to be binlinked'
    echo
    exit 1
fi
hab pkg exec jirotech/node_exporter/release-0.16/20180926010538 node_exporter "$@"
```

This happens for all binlinks EXCEPT for `core/hab` (which works like before), which must be linked for other binlinks to work.

These changes mean that runtime environment variables which are provided by Habitat are honored for binlinks (exec is the same as running a binlink).

Issues:
- This PR doesn't yet attempt to make Windows work
- This PR will cause binlinks to fail if `core/hab` is not binlinked / in $PATH
- `/hab/binlinks` is probably not the right place, happy to take advice

Signed-off-by: James Sewell <james.sewell@gmail.com>

### CODE REVIEW

**Comment by thesentinels** [comment]
Thanks for the pull request! Here is what will happen next:
 1. Your PR will be reviewed by the maintainers
 2. If everything looks good, one of them will `approve` it, and your PR will be merged.

Thank you for contributing!

File: components/hab/src/command/pkg/binlink.rs [hunk 1]
```diff
@@ -178,20 +193,57 @@ struct Binlink {
...
+    echo
+    exit 1 
+fi 
+hab pkg exec {0} {1} \"$@\"",
+            pkg_ident,
+            bin_name.to_string_lossy()
+        );
+
+        // Need to treat core/hab differently otherwise we get into a link loop
+        match format!("{}/{}", pkg_ident.origin, pkg_ident.name).as_ref() {
```
Comment (christophermaier) [inline components/hab/src/command/pkg/binlink.rs]: You shouldn't need to allocate a new string to check if this is `core/hab`... I'd just check if the origin was "core" and the name was "hab", and go from there.

File: components/hab/src/command/pkg/binlink.rs [hunk 2]
```diff
@@ -178,20 +193,57 @@ struct Binlink {
...
 
-        Ok(Self {
-            dest: dest_dir.as_ref().join(bin_name),
-            src: src.as_ref().to_path_buf(),
-        })
+        let template = format!(
+            "\
+#!/bin/sh 
+if ! [ -x \"$(command -v hab)\" ]; then 
+    echo 'ERROR: The core/hab package needs to be binlinked'
```
Comment (christophermaier) [inline components/hab/src/command/pkg/binlink.rs]: Doesn't necessarily need to be binlinked, just on `$PATH`, right?

File: components/eventsrv-protocol/src/message/event.rs [hunk 3]
```diff
@@ -2909,45 +2909,37 @@ impl ::protobuf::reflect::ProtobufValue for ServiceUpdate {
...
+    \x0c\x20\x01(\t\x12\x17\n\x0fupdate_strategy\x18\r\x20\x01(\t\x12\x13\n\
+    \x0bapplication\x18\x0e\x20\x01(\t\x12\x13\n\x0benvironment\x18\x0f\x20\
+    \x01(\t\x12\x0e\n\x06leader\x18\x10\x20\x01(\x08\x12\x10\n\x08follower\
+    \x18\x11\x20\x01(\x08\x12\x15\n\rupdate_leader\x18\x12\x20\x01(\x08\x12\
+    \x17\n\x0fupdate_follower\x18\x13\x20\x01(\x08\x12\x1b\n\x13election_is_\
+    running\x18\x14\x20\x01(\x08\x12\x1d\n\x15election_is_no_quorum\x18\x15\
+    \x20\x01(\x08\x12\x1c\n\x14election_is_finished\x18\x16\x20\x01(\x08\x12\
+    \"\n\x1aupdate_election_is_running\x18\x17\x20\x01(\x08\x12$\n\x1cupdate\
+    _election_is_no_quorum\x18\x18\x20\x01(\x08\x12#\n\x1bupdate_election_is\
+    _finished\x18\x19\x20\x01(\x08\
```
[the remark this answers is not in the document]
Comment (jamessewell) [reply components/eventsrv-protocol/src/message/event.rs]: Odd! Will remove

**Comment by mwrock** [comment]
Seems like we could bring the windows/linux implementations closer together now since this method comes much closer to the windows implementation.

**Comment by jamessewell** [comment]
> Seems like we could bring the windows/linux implementations closer together now since this method comes much closer to the windows implementation.

Yup! The only difference would be the stub payload (and what happens with core/hab I suppose)

File: components/hab/src/command/pkg/binlink.rs (hunk 2 above)
Comment (jamessewell) [reply components/hab/src/command/pkg/binlink.rs]: Fixed

File: components/hab/src/command/pkg/binlink.rs (hunk 1 above)
Comment (jamessewell) [reply components/hab/src/command/pkg/binlink.rs]: Ha, fixed. No idea why I chose this way of doing it. ☕️ needed maybe,

**Comment by jamessewell** [comment]
@fnichol Any comments before I start working on this some more?

**Comment by themightychris** [comment]
Shouldn't it use `exec` at the end?

**Comment by stevendanna** [comment]
Potentially:

Fixes #2376 
Fixes #4493

**Comment by raskchanky** [comment]
@jamessewell I'm so sorry that this PR has sat for as long as it has.  We're making an effort to review things in a more timely manner now, and it's my sincere hope that this kind of situation doesn't happen again.

After looking through the code here and reading the associated issues, specifically https://github.com/habitat-sh/habitat/issues/2376 and https://github.com/habitat-sh/habitat/issues/4493, I think this sounds like a good way forward.  If you wish to continue working on this, feel free, and I'd be happy to review the code when you're done.

**Comment by mwrock** [comment]
@jamessewell if you are tied up and not in a position where you can work on this soon, I'd like to carry the PR through if that is ok with you. I will of course preserve your commits.

**Comment by jamessewell** [comment]
Hi Matt,

Go for it! I was mainly waiting for someone to chime in on the way that
Windows should be handled.

Happy for you to take over.

Cheers,

**Comment by baumanj** [comment]
This might be the right time to address https://github.com/habitat-sh/habitat/issues/4493 as part of this.

**Comment by jamessewell** [comment]
This PR was created to address that issue!

**Comment by chefsalim** [comment]
A few comments - this direction/change looks promising, however, it also looks like it has the potential to be disruptive. Are there any backward compatibility implications?  If so, it would be good to have a plan for handling backward compatibility so that existing installations are not potentially inadvertently broken. It would be good to also have some test cases and/or a test plan + documentation. Finally, it might be worth considering putting this behind a feature flag or something similar so that it can be slowly eased in rather than all at once.

**Comment by themightychris** [comment]
Thinking about backwards compatibility implications, I can only think of three:

- If someone was previous working around the missing environment elements, their workaround might conflict with the issue being fixed
- If someone was parsing the target of binlink symlinks to read package version information, their matching may get broken by the changed path
- If someone was accidentally relying on binlinked commands having a missing environment, its introduction might cause issues

**Comment by baumanj** [comment]
@mwrock: can you let me know when you have this rebased against master? It's a bit tricky to review without being able to build, and I figured you'd need to rebase anyhow.

**Comment by mwrock** [comment]
I am closing this PR in favor of #6670. It preserves the commits of @jamessewell but can't share this PR since it is no longer in his fork.

**Pull request closed without merging**